### ISSUE TITLE
GORM: dataSource with empty dbCreate destroys DB tables and data 

### ISSUE THREAD

**Issue by hakanernam** [grails/grails-core#9410]
DataSource with empty dbCreate removes existing data from db in MySQL. Have to use some dummy value such as "donothing" to protect the data. 

Empty dbCreate also removes tables upon program exit via ^C. Example at:
https://github.com/hakanernam/directives.git

Tested in development mode:

|Run | dbCreate    | DB     | Result         |
|----|-------------|--------| -----------------------|
| 1  | "update"    | empty  | Tables created (OK)    |
| 2  | "update"    | full   | OK                     |
| 3  | "donothing" | empty  | Cannot find tables(OK) |
| 4  | "donothing" | full   | OK                     |
| 5  | empty       | empty  | Tables created         |
|    |             |        | Tables removed upon ^C |
| 6  | empty       | full   | Tables truncated (recreated?)      |
|    |             |        | Tables removed upon ^C |

**Comment by graemerocher** [comment]
The value is being read from one of the plugins you are using, the bug is the plugin

**Comment by hakanernam** [comment]
Thanks. This is a newly created app with only mysql driver added to build. I will file the issue for the plugin if you could tell me which one.

**Comment by Schlogen** [comment]
In your IDE do a search for `dbCreate`.

**Comment by hakanernam** [comment]
As per your suggestion I imported the project into Idea and did the search. Below is the result. All three hits are from application.yml. 

![screenshot at nov 12 20-20-37](https://cloud.githubusercontent.com/assets/1321157/11127081/f2b02f38-897a-11e5-873b-844c0e18765c.png)

**Comment by Schlogen** [comment]
You have to explicitly choose to search outside of the project

**Comment by hakanernam** [comment]
@Schlogen Thanks. I didnot know libraries could be searched like that. Never noticed the option.

It seems the value is coming from plugin.yml in Grails Cache Plugin

```
grails:
  profile: web-plugin
...
environments:
    development:
        dataSource:
            dbCreate: create-drop
```

I find code to load a default config it there is no config value, but of course a value comes from the plugin. So when I leave dbCreate value empty, it becomes "update" for production and test, "create-drop" for development. I guess a value of "do-nothing" will really keep ddl out of my db as a workaround.

**Comment by Schlogen** [comment]
@hakanernam If you find any plugins that have production dataSource configuration, please either let me know or submit a PR to remove that config

**Comment by graemerocher** [comment]
@Schlogen https://github.com/grails-plugins/grails-cache/blob/master/grails-app/conf/application.yml#L22

**Comment by hakanernam** [comment]
@Schlogen @graemerocher I also would like to suggest development default coming from plugins, if any, should be "update" rather than "create-drop". 

Development process may use duplicated (and quite large) real data and as such it is not possible to scaffold everytime during Bootstrap, so I believe being conservative should be the default. I have always set them empty and only turn them on during specific runs. 

And yes, I would be happy to try to do a PR. Pls let me know what you prefer.

**Comment by Schlogen** [comment]
@hakanernam I think you are in the minority there. I think the vast majority of people prefer `create-drop` when first laying out the domain structure of a new application, as it prevents manual work in dealing with the database.

**Comment by hakanernam** [comment]
@Schlogen You could be right. I just like to be explicit about it. It is not with every run of a new program that we deal with DB. My issue is not with the application profiles setting "create-drop" which I can see, but rather plugins doing it while I am not aware.

In any case, from now on I probably never could leave a dbCreate empty as I have been doing and will have to set to a dummy value  just to be safe.

**Comment by graemerocher** [comment]
I think maybe we should put a fix in core that makes it impossible to contribute data source config from a plugin.

**Comment by hakanernam** [comment]
@graemerocher Exactly what I would wish. 

@Schlogen Without disputing probably being in the minority about create-drop issue, in 
(https://github.com/grails/grails-data-mapping/blob/master/grails-datastore-gorm-hibernate4/src/main/groovy/grails/orm/bootstrap/HibernateDatastoreSpringInitializer.groovy)

```
 String defaultDataSourceBeanName = Mapping.DEFAULT_DATA_SOURCE
    String defaultSessionFactoryBeanName = SESSION_FACTORY_BEAN_NAME
    ***String ddlAuto = "update"***
    Set<String> dataSources = [defaultDataSourceBeanName]
```
and line 171

```
 def ddlAutoSetting = config.getProperty("${dsConfigPrefix}.dbCreate", ddlAuto)
```
leads me to believe the initial intention was also to set unspecified dbCreate values conservatively to "update".

**Comment by hakanernam** [comment]
@graemerocher I would like to curb my enthusiasm and retract my "wish". I have been thinking of delegating datasource configuration to an base inline-plugin for a while, inspired by spring-boot-starters,  and using it in other inline plugins and application itself as such. That probably would not make sense I in the fix you suggest.

**Comment by hakanernam** [comment]
I tried delegating datasource configuration to a base plugin of mine:

- Deleted application.yml in client plugin (or app)
- Duplicated delegate plugins's application.yml as plugin.yml (same directory)

I watched ddl creating "devDB". Apparently some other plugin wins the race to the keys. I found 3: "export", "cache" and one of my own. 

Apparently there is at this point no easy way to such delegation reliably. As long as arbitrary plugins can infect config space this way, I will have to keep a vigilant eye and explicitly set critical config properties. Also I believe plugin writers should review whether they unintentionally set any unnecessary properties just because it was in the default config file.

I can see it getting very tiresome, watching over every plugin I use. So it seems not allowing critical contributions from plugins may really be the safest way, unfortunately. 

I also realize this may have already been the situation before. Being just a blissfull grails user is over, I guess. :-(
